### ISSUE TITLE
Low GPU utilization

### ISSUE THREAD

**Issue by sshivam95** [dice-group/dice-embeddings#378]
I am training a dataset with the following stats:
```
Dataset file size on disk: 18G
Number of entities:179,706,494
Number of relations:15,219
Number of triples on train set:1,284,230,713
```
on a H100 GPU on [Otus cluster](https://upb-pc2.atlassian.net/wiki/spaces/PC2DOK/pages/601915506/Otus+Partitions) with the following configuration of the dice-embeddings:
```
dicee --path_single_kg "$data_path" \
        --model Keci \
        --num_epochs 500 \
        --p 0 \
        --q 1 \
        --embedding_dim 256 \
        --scoring_technique NegSample \
        --eval_model None \
        --batch_size 10_000 \
        --optim Adam \
        --trainer PL \
```
I want to address the low gpu utilization due to some non-parallel work on the cpu between epochs. The optimiser step is being conducted in the cpu. I have changed that so it can work on gpu using `fused=True`. However, here is the maximal GPU memory usage per GPU:
```
1 GPU : 23093MiB
2 GPUs: 13253MiB
4 GPUs: 9721MiB
6 GPUs: 8507MiB
```
Thus, it roughly scales as 5 GB+17.5 GB/NGPU.
The training time stays roughly the same, because the GPUs are only used to a very small extent in this case and most time is spent in a cpu core.

**Comment by Demirrr** [comment]
Thank you for providing a detailed description of the issue.

**Comment by Demirrr** [comment]
> I want to address the low gpu utilization due to some non-parallel work on the cpu between epochs. The optimiser step is being conducted in the cpu.

Something is wrong here. The optimiser step (updating the parameters) in the DDP cannot be conducted in the CPU.
See [this video of pytorch](https://www.youtube.com/watch?v=Cvdhwx-OBBo) as a refresher.

Yet, a mini-batch is generated on the CPU, see [the process of creating negative triples](https://github.com/dice-group/dice-embeddings/blob/c0c7a2963eb6251321a598f05c8ed40aace1f188/dicee/dataset_classes.py#L717).

**Comment by ashishtiwari03** [comment]
@Demirrr I agree the optimizer step itself should not be CPU-bound under DDP. I looked at the current NegSample path and the
  likely bottleneck seems to be `TriplePredictionDataset.collate_fn`, where the full positive+negative mini-batch is
  generated on CPU before transfer to GPU.

  I can work on a profiling-first PR that measures:
  - DataLoader/collate time
  - host-to-device transfer time
  - forward/backward/optimizer time

  If confirmed, I would try moving dynamic negative sampling to the training step on the target device, so the
  DataLoader only emits positive triples and corrupted triples/labels are created on GPU. I can also test
  `persistent_workers=True` for the DDP DataLoader.
